src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -171,31 +171,22 @@ class ChannelData {
                                       grpc_connectivity_state* state,
                                       grpc_closure* on_complete,
                                       grpc_closure* watcher_timer_init) {
-    auto* watcher = new ExternalConnectivityWatcher(
+    auto watcher = new ExternalConnectivityWatcher(
         this, pollent, state, on_complete, watcher_timer_init);
     {
       MutexLock lock(&external_watchers_mu_);
-      // Will be deleted when the watch is complete.
-      GPR_ASSERT(external_watchers_[on_complete] == nullptr);
-      external_watchers_[on_complete] = watcher;
+      // Store a ref to the watcher in the external_watchers_ map.
+      watcher->AddWatcherToExternalWatchersMapLocked(&external_watchers_,
```

Comment:
I don't think we need the `AddWatcherToExternalWatchersMapLocked()` method or the `Start()` method.  All of this work can be done in the ctor instead.  The ctor can explicitly acquite the lock while adding itself to the map.

Reply:
That seems cleaner. Thanks!